### DISCUSSION TITLE
Integration with ReactNative

### DISCUSSION THREAD

**Discussion by danielfttorres** [krzyzanowskim/ObjectivePGP#214] [General]
`.../Frameworks/ObjectivePGP.framework/strip-frameworks.sh: No such file or directory`

How I solve this?

**Comment by krzyzanowskim** [comment]
Just copy the file to the right place https://github.com/krzyzanowskim/ObjectivePGP/tree/master/scripts

![screen shot 2017-08-31 at 19 41 21](https://user-images.githubusercontent.com/758033/29936987-614b4602-8e84-11e7-82bd-99d7acc0cc81.png)

Looks like something went wrong with the automatic script installation. Will address that in next release.

**Comment by danielfttorres** [comment]
Can I replace this step using cocoa pods? I try to use cocoa pods, but receive this error: `Apple Mach-O Linker Error Group`. Any ideia? I try to link this project in React Native project.

**Comment by krzyzanowskim** [comment]
Cocoapods should work.
Not sure what `Apple Mach-O Linker Error Group` means. There should be more. I'm not a React Native expert, but I can try to build it by my own if you send me the sample project.

**Comment by danielfttorres** [comment]
Okey, thanks! The example is here: https://github.com/danielfeelfine/react-native-pgp-example

**Comment by krzyzanowskim** [comment]
It should build at all ?

```
/Users/marcinkrzyzanowski/Devel/react-native-pgp-example/node_modules/react-native/React/../third-party/glog-0.3.4/src/base/mutex.h:105:10: fatal error: 'config.h' file not found
#include "config.h"           // to figure out pthreads support
         ^
1 error generated.
```

**Comment by danielfttorres** [comment]
No. It's an another React Native error. You can solve this with:

In the Terminal, navigate to the `react-native/third-party/glog` folder inside node_modules (for me, this was `cd node_modules/react-native/third-party/glog-0.3.4`)
Once actively in this folder, `run ../../scripts/ios-configure-glog.sh`
Glog is configured and the required config.h header file is created for Xcode to find

Reference: https://github.com/facebook/react-native/issues/14382

**Comment by krzyzanowskim** [comment]
what about this:

```
** BUILD SUCCEEDED **


Installing build/Build/Products/Debug-iphonesimulator/PGPProject.app
An error was encountered processing the command (domain=NSPOSIXErrorDomain, code=2):
Failed to install the requested application
An application bundle was not found at the provided path.
Provide a valid path to the desired application bundle.
Print: Entry, ":CFBundleIdentifier", Does Not Exist

Command failed: /usr/libexec/PlistBuddy -c Print:CFBundleIdentifier build/Build/Products/Debug-iphonesimulator/PGPProject.app/Info.plist
Print: Entry, ":CFBundleIdentifier", Does Not Exist
```

**Comment by danielfttorres** [comment]
Select a developer team in General settings of project, ex.:

![screen shot 2017-08-31 at 15 48 23](https://user-images.githubusercontent.com/1668289/29939979-1a1d3904-8e64-11e7-9e71-aa45556ef532.png)

**Comment by krzyzanowskim** [comment]
that's not that. Info.plist is not at the location the script think it is.

**Comment by danielfttorres** [comment]
It's about the first error or about this latest error? In the latest error, you must be select a team and solve the problem.

**Comment by krzyzanowskim** [comment]
it's that https://github.com/facebook/react-native/issues/10102
the script can't correctly setup path to ios binaries

basically this call
```
Building using "xcodebuild -workspace PGPProject.xcworkspace -configuration Debug -scheme PGPProject -destination id=F37CCCE2-7942-4EA7-9088-D5D117FE5313 -derivedDataPath build"
User defaults from command line:
    IDEDerivedDataPathOverride = /Users/marcinkrzyzanowski/react-native-pgp-example/ios/build
```

override derived data path, to non-existing

**Comment by danielfttorres** [comment]
Were you able to compile successfully?

**Comment by krzyzanowskim** [comment]
ok, figured out. Needed to edit `runIOS.js` to use correct paths. Where is the problem ?
![screen shot 2017-08-31 at 21 15 14](https://user-images.githubusercontent.com/758033/29941072-7dc0d9c0-8e91-11e7-8af4-74b433acbf38.png)

**Comment by danielfttorres** [comment]
The problem is `strip-frameworks.sh: No such file or directory` when compiling after install *ObjectivePGP* using CocoaPods. Did you use CocoaPods?

**Comment by krzyzanowskim** [comment]
yes, I used the project you created, fixed react-native and build. Here is the full archive:
https://drive.google.com/open?id=0BzncA1vRoAW_d0ZObVlTUzNvQUU

**Comment by danielfttorres** [comment]
How you fixed react-native?

**Comment by krzyzanowskim** [comment]
I had to patch `runIOS.js` script from react-native to use proper paths.

**Comment by danielfttorres** [comment]
Your project is different from my. My code use ObjectivePGP and your, not. So it worked (I think). Try installing ObjectivePGP and adding the files:

*ReactNativePGP.m*
```
#import "ReactNativePGP.h"
#import <React/RCTLog.h>
//#include <ObjectivePGP/ObjectivePGP.h>

@implementation ReactNativePGP

//ObjectivePGP *pgp = [[ObjectivePGP alloc] init];

// To export a module named ReactNativePGP
RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(addEvent:(NSString *)name location:(NSString *)location)
{
  RCTLogInfo(@"Pretending to create an event %@ at %@", name, location);
}

@end
```

*ReactNativePGP.h*
```
//
//  ReactNativePGP.h
//  PGPProject
//
//  Created by Daniel Torres on 31/08/17.
//  Copyright © 2017 Facebook. All rights reserved.
//

#import <React/RCTBridgeModule.h>

@interface ReactNativePGP : NSObject <RCTBridgeModule>
@end
```

**Comment by krzyzanowskim** [comment]
those files are not part of the Xcode project, how should it work? The cocoapods installed the ObjectivePGP.framework though.

**Comment by danielfttorres** [comment]
How can I add to my XCode project?

You'd better start a fresh react-native (it's fast) project and try to adding ObjectivePGP in it. And so you could reproduce the problem.

**Comment by krzyzanowskim** [comment]
> How can I add to my XCode project?

drag file to xcode and add to target

> You'd better start a fresh react-native (it's fast)

I rather not, it doesn't look like the toolchain is working out of the box.

**Comment by krzyzanowskim** [comment]
anyway, the cocoapods setup works with ObjectivePGP, here is the current version of the project I build:

https://github.com/krzyzanowskim/react-native-pgp-example

**Comment by danielfttorres** [comment]
When try to run your fork, I take the same error (commented in https://github.com/krzyzanowskim/ObjectivePGP/issues/70#issuecomment-326370907):

![screen shot 2017-08-31 at 18 27 28](https://user-images.githubusercontent.com/1668289/29946114-7187f024-8e7a-11e7-9236-253a272d39ca.png)

**Comment by krzyzanowskim** [comment]
What's your Xcode version?

![screen shot 2017-08-31 at 23 31 52](https://user-images.githubusercontent.com/758033/29946181-98f32ac8-8ea4-11e7-8727-0730259310e7.png)

**Comment by danielfttorres** [comment]
Version 8.3.3 (8E3004b) too.

**Comment by krzyzanowskim** [comment]
¯\_(ツ)_/¯

 I'm running with this command:
`react-native run-ios`
that build the app for the simulator.

Logs: https://gist.githubusercontent.com/krzyzanowskim/eb372bb658215021da3f919cb2ba79c6/raw/95e25bce0cf0d4ac029d3aa05e251baf4eae1219/rn.log

**Comment by danielfttorres** [comment]
Thanks! I `BUILD SUCCEEDED` too by run `react-native run-ios` instead of `CMD+B` in Xcode.

Now I will try to expose the ObjetivePGP methods using iOS [Native Modules](https://facebook.github.io/react-native/docs/native-modules-ios.html) from react-native, but first, I have some questions:

1. Can I generate a pair of keys and store that into device with OBjetivePGP? How do you think it would be safer? What method I use to this?
2. How long does a device take to generate a key? 1-2 seconds? And how long does it take to sign a message?
3. I can decrypt key and sign a message with that key using ObjetivePGP? What methods I use to this?

**Comment by krzyzanowskim** [comment]
> Can I generate a pair of keys and store that into device with OBjetivePGP? How do you think it would be safer? What method I use to this?

not today. This feature will be available with next update. It's currently WIP.

> I can decrypt key and sign a message with that key using ObjetivePGP? What methods I use to this?

If I understand the question right, you just sign with password https://github.com/krzyzanowskim/ObjectivePGP#sign-data-or-file

**Comment by danielfttorres** [comment]
Do you know if there is any module that generates keys in iOS? Would it be a difficult task to do one or implement this method?

I'm thinking of using: https://github.com/kryptco/kryptonite-ios. What do you think?

Thanks

**Comment by krzyzanowskim** [comment]
as I said, key generation is in progress ETA in 2 weeks. I'm rather not going to verify the kryptonite compliance.

**Comment by danielfttorres** [comment]
Great @krzyzanowskim. I will wait for the implementation. In the meantime, I'll be working on the RN wrapper of signature methods.

Thanks
